### PR TITLE
Add follow_inodes parameter on in_tail

### PR DESCRIPTION
**Pull request by cosmo0920**
**Which issue(s) this PR fixes**: 
Revive of #1660  

**What this PR does / why we need it**: 
From #1660:

> Add ability to track inodes inside fluentd in_tail plugin to enable combination of '*' in `path` configuration with rotation inside same directory and `read_from_head` set to true. In proposed solution we add `follow_inodes` flag (set to `false` by default to preserve current behavior). When `follow_inodes` is set to `true` and file rotation is detected - we do not reread file, instead we are checking if such inode is present inside pos file - if yes, we don't recreate TailWatcher. TailWatcher is closed after `rotate_wait` period and file marked as unwatched inside position file after `limit_recently_modified` time passed. Pos file is cleaned on startup as it was before, but also delete entries for non-existing files.

And we also think that this following inodes feature is useful for tracking k8s related logs.

**Docs Changes**:
Needed for `follow_inodes` parameter. I'll register a PR for it.

**Release Note**: 

Same as title.

### CODE REVIEW

File: lib/fluent/plugin/in_tail.rb [hunk 1]
```diff
@@ -312,21 +341,21 @@ def expand_paths
...
 
-      unwatched = existence_paths - target_paths
-      added = target_paths - existence_paths
+      unwatched_hash = existence_paths_hash.reject {|key, value| target_paths_hash.key?(key)}
+      added_hash = target_paths_hash.reject {|key, value| existence_paths_hash.key?(key)}
 
-      stop_watchers(unwatched, immediate: false, unwatched: true) unless unwatched.empty?
-      start_watchers(added) unless added.empty?
+      stop_watchers(unwatched_hash.values, immediate: false, unwatched: true) unless unwatched_hash.empty?
+      start_watchers(added_hash.values) unless added_hash.empty?
```
Comment (repeatedly) [inline lib/fluent/plugin/in_tail.rb:353]: Calling `values` is needed? Passes `Hash` and call `each_value` has a problem? Latter approach doesn't create temporal object.

File: lib/fluent/plugin/in_tail.rb [hunk 2]
```diff
@@ -303,7 +308,31 @@ def expand_paths
...
+        else
+          hash[tuple.path] = tuple
+        end
+      }
+      hash
+    end
+
+    def existence_path
+      hash = {}
+      @tails.keys.each {|path_ino|
```
Comment (repeatedly) [inline lib/fluent/plugin/in_tail.rb:328]: How about `each_key` ?

File: lib/fluent/plugin/in_tail/position_file.rb [hunk 3]
```diff
@@ -224,5 +263,7 @@ def read_inode
         @inode
       end
     end
+
+    PathInodeTuple = Struct.new(:path, :ino)
```
Comment (repeatedly) [inline lib/fluent/plugin/in_tail/position_file.rb:267]: What is the differences from `PathWithInode`?

Comment (cosmo0920) [reply lib/fluent/plugin/in_tail/position_file.rb:258]: They are same behavior except for member names.
`PathWithInode` is not used anywhere. It should be removed the PR.

File: lib/fluent/plugin/in_tail.rb (hunk 1 above)
Comment (cosmo0920) [reply lib/fluent/plugin/in_tail.rb]: Fixed in https://github.com/fluent/fluentd/pull/3182/commits/e294104fa28b1ed6dca2c7eda7a105834493a1b5.

File: lib/fluent/plugin/in_tail.rb (hunk 2 above)
Comment (cosmo0920) [reply lib/fluent/plugin/in_tail.rb]: Fixed in https://github.com/fluent/fluentd/pull/3182/commits/d8019edd6c3625bd7e0fda7f5109becaa4909ac1.

File: lib/fluent/plugin/in_tail/position_file.rb (hunk 3 above)
Comment (cosmo0920) [reply lib/fluent/plugin/in_tail/position_file.rb:258]: Fixed in https://github.com/fluent/fluentd/pull/3182/commits/c1f41da80dbf9ff6289bb01031a956a3b7dc1cec.

File: lib/fluent/plugin/in_tail.rb [hunk 4]
```diff
@@ -608,10 +657,12 @@ def on_timer
     end
 
     class TailWatcher
-      def initialize(path, pe, log, read_from_head, update_watcher, line_buffer_timer_flusher, io_handler_build)
+      def initialize(path, ino, pe, log, read_from_head, follow_inodes, update_watcher, line_buffer_timer_flusher, io_handler_build)
```
Comment (repeatedly) [inline lib/fluent/plugin/in_tail.rb:660]: `path` and `ino` can be merged into `PathInodeTuple`?
It also reduces method argument in other places, e.g `unwatch`, `setup_watcher` and others.

Of course, separate PR is ok for me.

File: lib/fluent/plugin/in_tail/position_file.rb [hunk 5]
```diff
@@ -224,5 +254,7 @@ def read_inode
...
```
Comment (repeatedly) [inline lib/fluent/plugin/in_tail/position_file.rb:258]: I'm re-consindering this struct name. This name depends on members.
If we add more member in the future, this name will be invalid.
So `TargetInfo` or other name is better for the future.

Comment (cosmo0920) [reply lib/fluent/plugin/in_tail/position_file.rb]: Fixed in https://github.com/fluent/fluentd/pull/3182/commits/72b9d4c6c7e0505e51da2f1f5024056bf71ef6c7.

File: lib/fluent/plugin/in_tail.rb (hunk 4 above)
Comment (cosmo0920) [reply lib/fluent/plugin/in_tail.rb:660]: Oh, that's reasonable. Will do.

Comment (cosmo0920) [reply lib/fluent/plugin/in_tail.rb]: Fixed in https://github.com/fluent/fluentd/pull/3182/commits/08ee2da9c3ac895b747dbabd2ef2a0cada1bcc25.

Comment (cosmo0920) [reply lib/fluent/plugin/in_tail.rb]: Ah, wait. Merging path and inode arguments makes slight a bit unstable for testing...
Perhaps we should do in the separated PR.

**Review by repeatedly — APPROVED** [review]
LGTM. 

There are several refactoring points.

- Some code still use `path_with_inode` name
- Some methods still take two `path`/`ino` arguments. It is not needed because we can use TargetInfo or TailWatcher instance for it, TailInput methods, TailWatcher methods, PositionFile methods and more. If we need to keep this two argument version, e.g. for path and different inode check, adding new TargetInfo version is better for avoiding `func(target_info.path, target_info.ino)` like redundant code.

New PR is better for large refactoring.

**Comment by cosmo0920** [comment]
OK, let's merge this PR.
Refactoring should do in the another PR.

**Pull request merged**